### PR TITLE
replace implementation by compose-on-kubernetes

### PR DESCRIPTION
**Pull request by jdrouet**
Signed-off-by: Jérémie Drouet <jeremie.drouet@gmail.com>

**- What I did**
Replaced the implementation by the compose-on-kubernetes implementation

**- How I did it**
- Integrated the changes made in the cli repository in the compose-on-kubernetes repository
- Alias the `/kubernetes` folder to the compose-on-kubernetes implementation
- Added the compose on kubernetes dependency
- Replace the import to `/kubernetes` by compose-on-kubernetes in the `cli` package

**- How to verify it**
Deploy a stack on kubernetes with docker cli

**- Description for the changelog**
Replace the kubernetes folder implementing for the deployment of compose files on kubernetes by the implementation made in compose-on-kubernetes

**- A picture of a cute animal (not mandatory but encouraged)**
image

### CODE REVIEW

File: vendor.conf [hunk 1]
```diff
@@ -13,6 +13,7 @@ github.com/davecgh/go-spew 346938d642f2ec3594ed81d874461961cd0faa76 # v1.1.0
 github.com/dgrijalva/jwt-go a2c85815a77d0f951e33ba4db5ae93629a1530af
 github.com/docker/distribution 83389a148052d74ac602f5f1d62f86ff2f3c4aa5
 github.com/docker/docker 6e3113f700dea1bf2785d94731b4b5a1e602d9ab
+github.com/docker/compose-on-kubernetes 60262c4a8b724e82b940fecf587eac18f00a9738 https://github.com/jdrouet/compose-on-kubernetes.git
```
Comment (vdemeester) [inline vendor.conf]: We're gonna need this to be up-to-date and merged on master :angel:

File: kubernetes/client/clientset/clientset.go [hunk 2]
```diff
@@ -1,96 +1,23 @@
 package clientset
 
 import (
-	composev1beta1 "github.com/docker/cli/kubernetes/client/clientset/typed/compose/v1beta1"
-	composev1beta2 "github.com/docker/cli/kubernetes/client/clientset/typed/compose/v1beta2"
-	"github.com/golang/glog"
-	"k8s.io/client-go/discovery"
-	"k8s.io/client-go/rest"
-	"k8s.io/client-go/util/flowcontrol"
+	api "github.com/docker/compose-on-kubernetes/api/client/clientset"
```
Comment (thaJeztah) [inline kubernetes/client/clientset/clientset.go]: WIth these now being a single import, we should probably remove the brackets, and change them to

```suggestion
import api "github.com/docker/compose-on-kubernetes/api/client/clientset"
```

File: kubernetes/client/clientset/clientset.go [hunk 3]
```diff
@@ -1,96 +1,23 @@
...
 )
 
 // Interface defines the methods a compose kube client should have
 // FIXME(vdemeester) is it required ?
```
Comment (thaJeztah) [inline kubernetes/client/clientset/clientset.go]: These `FIXME`'s can be removed, as they'll have to be done in the new location now

File: kubernetes/client/clientset/clientset.go [hunk 4]
```diff
@@ -1,96 +1,23 @@
...
-type Interface interface {
-	Discovery() discovery.DiscoveryInterface
-	ComposeV1beta2() composev1beta2.ComposeV1beta2Interface
-	ComposeV1beta1() composev1beta1.ComposeV1beta1Interface
-}
+type Interface = api.Interface
```
Comment (thaJeztah) [inline kubernetes/client/clientset/clientset.go]: To encourage people to move to the new package, might want to officially mark them deprecated by adding a deprecation comment.

Not sure what the best description would be to clearly direct people to the right location (@vdemeester any suggestions?). Perhaps something like;

```suggestion
// Deprecated: Use github.com/docker/compose-on-kubernetes/api/client/clientset.Interface instead
type Interface = api.Interface
```

(Same for all the other aliases we created)

File: vendor.conf [hunk 5]
```diff
@@ -1,3 +1,4 @@
+cloud.google.com/go ac3720150beed4ebd66918aa31d4713110098260
```
Comment (thaJeztah) [inline vendor.conf]: I see they tag releases https://github.com/googleapis/google-cloud-go/releases; can we use the commit of a tagged version for this, and add a comment to mention that version? e.g.

```suggestion
cloud.google.com/go 0ebda48a7f143b1cce9eb37a8c1106ac762a3430 # v0.34.0
```

(current version looks to be take directly from master, and is 29 commits ahead of v0.34.0 https://github.com/googleapis/google-cloud-go/compare/v0.34.0...ac3720150beed4ebd66918aa31d4713110098260)

File: vendor.conf [hunk 6]
```diff
@@ -13,6 +14,7 @@ github.com/davecgh/go-spew 346938d642f2ec3594ed81d874461961cd0faa76 # v1.1.0
 github.com/dgrijalva/jwt-go a2c85815a77d0f951e33ba4db5ae93629a1530af
 github.com/docker/distribution 83389a148052d74ac602f5f1d62f86ff2f3c4aa5
 github.com/docker/docker 6e3113f700dea1bf2785d94731b4b5a1e602d9ab
+github.com/docker/compose-on-kubernetes 939a5ba6c6dbfc78e6cc29b612716d5688b9a323
```
Comment (thaJeztah) [inline vendor.conf]: Will we create a release for this? (I see there's a v0.4.16 tag https://github.com/docker/compose-on-kubernetes/releases but this PR probably needs changes that were merged after this https://github.com/docker/compose-on-kubernetes/compare/v0.4.16...939a5ba6c6dbfc78e6cc29b612716d5688b9a323)

File: vendor.conf (hunk 5 above)
Comment (thaJeztah) [reply vendor.conf]: Actually; I see compose-on-kubernetes uses that version as well https://github.com/docker/compose-on-kubernetes/commit/be6d5be2f61bdfdc0716a5c50393e4f4720e8800#diff-bd247e83efc3c45ae9e8c47233249f18R18

File: vendor.conf [hunk 7]
```diff
@@ -83,6 +85,7 @@ github.com/xeipuuv/gojsonreference bd5ef7bd5415a7ac448318e64f11a24cd21e594b
 github.com/xeipuuv/gojsonschema 93e72a773fade158921402d6a24c819b48aba29d
 golang.org/x/crypto 0709b304e793a5edb4a2c0145f281ecdc20838a4
 golang.org/x/net a680a1efc54dd51c040b3b5ce4939ea3cf2ea0d1
+golang.org/x/oauth2 d668ce993890a79bda886613ee587a69dd5da7a6
```
Comment (thaJeztah) [inline vendor.conf]: Perhaps we should make vendoring match the upstream kubernetes-on-compose version? https://github.com/docker/compose-on-kubernetes/blob/939a5ba6c6dbfc78e6cc29b612716d5688b9a323/Gopkg.lock#L725

That way we know we're using the same version as the tests on kubernetes-on-compose run on


Might be worth checking if kubernetes-on-compose is using more recent versions of other packages as well

File: kubernetes/client/clientset/clientset.go (hunk 4 above)
Comment (vdemeester) [reply kubernetes/client/clientset/clientset.go]: @thaJeztah yes that works, most of the IDE/editor should pick those (and mark them as deprecated in some way)

File: vendor.conf (hunk 6 above)
Comment (silvin-lubecki) [reply vendor.conf]: We will release a new version today yes, and then modify the sha to target it, thanks for catching that @thaJeztah 😄

File: kubernetes/check.go [hunk 8]
```diff
@@ -1,55 +1,17 @@
...
 
 // StackVersion represents the detected Compose Component on Kubernetes side.
-type StackVersion string
+// Deprecated: Use github.com/docker/compose-on-kubernetes/api.StackVersion instead
+type StackVersion = api.StackVersion
 
 const (
 	// StackAPIV1Beta1 is returned if it's the most recent version available.
-	StackAPIV1Beta1 = StackVersion("v1beta1")
+	StackAPIV1Beta1 = api.StackAPIV1Beta1
```
Comment (thaJeztah) [inline kubernetes/check.go]: Should likely have `Deprecated:` as well 😅

File: kubernetes/check.go [hunk 9]
```diff
@@ -1,55 +1,17 @@
...
 	// StackAPIV1Beta2 is returned if it's the most recent version available.
-	StackAPIV1Beta2 = StackVersion("v1beta2")
+	StackAPIV1Beta2 = api.StackAPIV1Beta2
```
Comment (thaJeztah) [inline kubernetes/check.go]: same here

File: kubernetes/check.go [hunk 10]
```diff
@@ -1,55 +1,17 @@
...
-			for _, version := range group.Versions {
-				if version.Version == stackAPI.Version {
-					return true
-				}
-			}
-		}
-	}
-	return false
-}
+var GetStackAPIVersion = api.GetStackAPIVersion
```
Comment (thaJeztah) [inline kubernetes/check.go]: Same here, and some others below (sorry 😊 I realize it's a bit of work )

File: vendor.conf [hunk 11]
```diff
@@ -13,10 +14,11 @@ github.com/davecgh/go-spew 346938d642f2ec3594ed81d874461961cd0faa76 # v1.1.0
 github.com/dgrijalva/jwt-go a2c85815a77d0f951e33ba4db5ae93629a1530af
 github.com/docker/distribution 83389a148052d74ac602f5f1d62f86ff2f3c4aa5
 github.com/docker/docker 6e3113f700dea1bf2785d94731b4b5a1e602d9ab
+github.com/docker/compose-on-kubernetes 939a5ba6c6dbfc78e6cc29b612716d5688b9a323
 github.com/docker/docker-credential-helpers 5241b46610f2491efdf9d1c85f1ddf5b02f6d962
 # the docker/go package contains a customized version of canonical/json
 # and is used by Notary. The package is periodically rebased on current Go versions.
-github.com/docker/go d30aec9fd63c35133f8f79c3412ad91a3b08be06
+github.com/docker/go 0ebda48a7f143b1cce9eb37a8c1106ac762a3430 # v0.34.0
```
Comment (thaJeztah) [inline vendor.conf]: Looks like you updated the wrong line here;  this tag/commit is for `cloud.google.com/go`

Probably that caused the huge diff in this commit

screen shot 2018-12-20 at 14 30 17

**Comment by silvin-lubecki** [comment]
@thaJeztah PTAL 😄

**Comment by silvin-lubecki** [comment]
@thaJeztah PTAL I think @jdrouet addressed all your comments  👍

File: vendor.conf (hunk 5 above)
Comment (thaJeztah) [reply vendor.conf]: Looks like you forgot to update this to the version (v0.3.4) that's used by compose-on-kubernetes ( see my [earlier comment](https://github.com/docker/cli/pull/1583/files#r243203663); https://github.com/docker/compose-on-kubernetes/blob/6347392617e1efd3077f8a918d67fa94cd39289e/Gopkg.lock#L12-L18

**Comment by jdrouet** [comment]
@thaJeztah it's done 😉

**Comment by thaJeztah** [comment]
Whoop, thanks!

**Pull request merged**